apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -87,76 +81,154 @@ export function positionBlocksOnWorkspace(workspace, blockOrderMap) {
+ * Such that none of the blocks overlap with each other
+ * @param {Array<Blockly.Block>} blocks - The blocks to position
+ * @param {Blockly.Workspace} workspace - The current Blockly workspace
+ */
+function adjustBlockPositions(blocks, workspace) {
+  const {contentWidth = 0, viewWidth = 0} = workspace.getMetrics();
+  const {RTL} = workspace;
+
+  const padding = viewWidth ? WORKSPACE_PADDING : 0;
+  const width = viewWidth || contentWidth;
+
+  // If the workspace is RTL, horizontally mirror the starting position
+  let x = RTL ? width - padding : padding;
+  let y = padding;
+
+  let existingColliders = [];
+  let blocksToPlace = [];
+  blocks.forEach(block => {
+    if (isBlockLocationUnset(block)) {
+      blocksToPlace.push(block);
+    } else {
+      insertCollider(existingColliders, getCollider(block));
+    }
+  });
+
+  blocksToPlace.forEach(block => {
+    let collider = getCollider(block);
+    existingColliders.forEach(existingCollider => {
+      if (isOverlapping(collider, existingCollider)) {
+        y =
+          existingCollider.y +
+          existingCollider.height +
+          VERTICAL_SPACE_BETWEEN_BLOCKS +
+          SVG_FRAME_TOP_PADDING;
+        block.moveTo({x, y});
+        // Collision area must be updated to account for new position
+        collider = getCollider(block);
+      }
+    });
+    insertCollider(existingColliders, collider);
   });
 }
 
 /**
- * Use a cursor to position a block on a workspace (if it does not already have a position)
- * @param {Blockly.Block} block - the block to be moved
- * @param {object} cursor - a location for moving a block, if needed
- * @param {number} cursor.x - an x-coordinate for moving a block
- * @param {number} cursor.y - a y-coordinate for moving a block
- * @return {object} the cursor with updated coordinates
+ * A "collider" is a an object that contains coordinates and dimensions of a block, adjusted
+ * for anything that affects the area of the workspace it occupies (for now, just the SVG frame)
+ * @typedef {Object} Collider
+ * @property {number} x - The x-coordinate of the block, adjusted for SVG frame padding
+ * @property {number} y - The y-coordinate of the block, adjusted for SVG frame padding
+ * @property {number} height - The height of the block, including the SVG frame height
+ * @property {number} width - The width of the block, accounting for SVG frame width on either side
  */
-function positionBlockWithCursor(block, cursor) {
-  if (isBlockLocationUnset(block)) {
-    block.moveTo(getNewLocation(block, cursor));
-    cursor.y += getCursorYAdjustment(block);
+function getCollider(block) {
+  const position = block.getRelativeToSurfaceXY();
+  const size = block.getHeightWidth();
+
+  const collider = {
+    ...position,
+    ...size,
+  };
+
+  // SVG frames require us to account for additional height and width
+  // TODO: How do we handle RTL here?
+  if (block.functionalSvg_) {
+    collider.x -= SVG_FRAME_SIDE_PADDING;
```

Comment:
@mikeharv How do we need to adjust this for RTL?

Reply:
From Mike via Slack:
> the position is also consistent in RTL mode, which is great news. if we add the horizontal adjustment for frames, i think we’d just want to subtract the same amount for RTL.

Reply:
Tested in RTL - looks to be working well!